### PR TITLE
Clarify state of transcript hash through different types

### PR DESCRIPTION
**Pull request by djc**
Eliminates 2 unwrap calls in the `HandshakeHash` module.

### CODE REVIEW

File: rustls/src/client/tls12.rs [hunk 1]
```diff
@@ -490,7 +490,14 @@ fn emit_certverify(
         Some(signer) => signer,
     };
 
-    let message = transcript.take_handshake_buf();
+    let message = match transcript.take_handshake_buf() {
+        Some(msg) => msg,
+        None => {
+            trace!("Not sending CertificateVerify, no transcript");
+            return Ok(());
+        }
```
Comment (ctz) [inline rustls/src/client/tls12.rs:498]: Can we match on both client_auth.signer.take() and transcript.take_handshake_buf() at once?

File: rustls/src/server/tls12.rs [hunk 2]
```diff
@@ -661,12 +661,20 @@ impl hs::State for ExpectCertificateVerify {
...
+                Some(msgs) => {
+                    let certs = &self.client_cert;
+                    self.config
+                        .verifier
+                        .verify_tls12_signature(&msgs, &certs[0], sig)
+                }
+                None => {
+                    cx.common
+                        .send_fatal_alert(AlertDescription::AccessDenied);
+                    Err(Error::General("client authentication not set up".into()))
```
Comment (ctz) [inline rustls/src/server/tls12.rs:675]: if this is reachable, can it have a test?

File: rustls/src/client/tls12.rs (hunk 1 above)
Comment (ctz) [reply rustls/src/client/tls12.rs:498]: (to achieve the aim of one piece of code that handles the not-sending CertificateVerify case)

Comment (djc) [reply rustls/src/client/tls12.rs]: Yup, done.

Comment (djc) [reply rustls/src/client/tls12.rs]: (This has two different takes on not-sending, let me know if you want to collapse those.)

File: rustls/src/server/tls12.rs (hunk 2 above)
Comment (djc) [reply rustls/src/server/tls12.rs:680]: I don't think this is actually reachable:

* We can only get to the `ExpectCertificateVerify` state if `ExpectClientKx::client_cert` is `Some`
* An `ExpectClientKx` state with `client_cert` set to `Some` is only created by `ExpectCertificate`
* In `ExpectCertificate`, we might abandon client auth, but we set `ExpectClientKx::client_cert` to `None` if so
* `ExpectCertificate` is set up only if we're doing client auth, meaning `config.verifier.offer_client_auth()` is `true`
* In `ExpectClientHello::new()`, we initialized the `HandshakeHashBuffer` with client auth enabled under the same condition

I've added a comment with a condensed version of  this reasoning.

**Comment by ctz** [comment]
Thanks!

**Pull request merged**